docs/tokencache.md

```diff
@@ -2,16 +2,350 @@
+
+ID Token & Client_Info -> Account & Tenant Profile
+Access Token -> Credential
+Refresh Token -> Credential
+
+## Universal Schema at a high level / OAuth v2 Mapping
+
+- Accounts
+  - ID Token of account home tenant
+- Tenant Profiles - The representation of 1 account in 1 or more tenants (external account guest/member scenarios)
+  - ID Token of the account home tenant
+  - ID Token from other tenants that the account is a member of
+- Credentials
+  - Id Token
+  - Access Tokens
+  - Refresh Tokens
+  - etc....
+
+### Relationships
+
+1 Account has 1 or more credentials (minimum, the `AccountRecord` + `RefreshTokenRecord`)
+1 Credential belongs to 1 account
+
+## OAuth 2 Token Cache
+
+This section describes the base class for the higher level of the TokenCache.  The level responsible for translated an oAuth response into universal cache schema artifacts.
+
+There are different implementations of the cache based on the library using it and the scenarios it needs to support.
+
+> NOTE: I'm going to ignore the ADALOAuth2TokenCache in this document. It deals with interacting with the legacy ADAL token cache.
+
+### Abstract base class
+
+The abstract base class for all caches is the OAuth2TokenCache.  The physical schema of the cache in terms of account and credential records is decoupled from the protocol used to acquire the account and credential; however the cache interface knows about oAuth2 token responses and how to convert those to accounts and crentials and persist them to storage.
+
+```java
+//Example
+/**
+     * Saves the credentials and tokens returned by the service to the cache.
+     *
+     * @param oAuth2Strategy The strategy used to create the token request.
+     * @param request        The request used to acquire tokens and credentials.
+     * @param response       The response received from the IdP/STS.
+     * @return The {@link ICacheRecord} containing the Account + Credentials saved to the cache.
+     * @throws ClientException If tokens cannot be successfully saved.
+     */
+    public abstract ICacheRecord save(final T oAuth2Strategy,
+                                      final U request,
+                                      final V response) throws ClientException;
+```
+
+### MSALOAuth2TokenCache
+
+The simplest version of the cache is the MSALOAuth2TokenCache.  There is one instance of this cache per PublicClientApplication created using MSAL within a client android application.  The important thing to note is that there is 1 cache per public client application identified by it's client id.
+
+### Family of Client Ids (Special Case)
+
+Applications published by Microsoft may be authorized to share refresh tokens with one another.  All applications that are authorized to share tokens share a single MSALOAuth2TokenCache instance.  They effectively share a client id, which in this case is the family id (generally 1 is used to represent the family)
```

Comment:
How are we currently verifying if an App B is authorized to use the FRT populated by App A? I could find the ESTS call to fetch the token with FRT and verify that it is authorized to use it.  Do we have to make this network call every time for a new app?

Reply:
AndroidBrokerPlatformComponents.java in the broker repo has an isAuthorizedToShareTokens method that you may want to review.